graph/src/components/adapter.rs

```diff
@@ -0,0 +1,762 @@
+    blockchain::{BlockHash, ChainIdentifier},
+    cheap_clone::CheapClone,
+    data::value::Word,
+    prelude::error,
+    tokio::sync::RwLock,
+};
+
+use crate::components::store::{BlockStore as BlockStoreTrait, ChainStore as ChainStoreTrait};
+
+const VALIDATION_ATTEMPT_TTL_SECONDS: i64 = 60 * 5;
+
+#[derive(Debug, Error)]
+pub enum ProviderManagerError {
+    #[error("unknown error {0}")]
+    Unknown(#[from] anyhow::Error),
+    #[error("provider {provider} failed verification, expected ident {expected}, got {actual}")]
+    ProviderFailedValidation {
+        provider: ProviderName,
+        expected: ChainIdentifier,
+        actual: ChainIdentifier,
+    },
+    #[error("no providers available for chain {0}")]
+    NoProvidersAvailable(ChainId),
+    #[error("all providers for chain_id {0} have failed")]
+    AllProvidersFailed(ChainId),
+}
+
+#[async_trait]
+pub trait NetIdentifiable: Sync + Send {
+    async fn net_identifiers(&self) -> Result<ChainIdentifier, anyhow::Error>;
+    fn provider_name(&self) -> ProviderName;
+}
+
+#[async_trait]
+impl<T: NetIdentifiable> NetIdentifiable for Arc<T> {
+    async fn net_identifiers(&self) -> Result<ChainIdentifier, anyhow::Error> {
+        self.as_ref().net_identifiers().await
+    }
+    fn provider_name(&self) -> ProviderName {
+        self.as_ref().provider_name()
+    }
+}
+
+pub type ProviderName = Word;
+pub type ChainId = Word;
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
+struct Ident {
+    provider: ProviderName,
+    chain_id: ChainId,
+}
+
+#[derive(Error, Debug)]
+pub enum IdentValidatorError {
+    #[error("Store ident wasn't set")]
+    UnsetIdent,
+    #[error("the net version for chain {chain_id} has changed from {store_net_version} to {chain_net_version} since the last time we ran")]
+    ChangedNetVersion {
+        chain_id: ChainId,
+        store_net_version: String,
+        chain_net_version: String,
+    },
+    #[error("the genesis block hash for chain {chain_id} has changed from {store_hash} to {chain_hash} since the last time we ran")]
+    ChangedHash {
+        chain_id: ChainId,
+        store_hash: BlockHash,
+        chain_hash: BlockHash,
+    },
+    #[error("unable to get store for chain {0}")]
+    UnavailableStore(ChainId),
+}
+
+#[async_trait]
+pub trait IdentValidator: Sync + Send {
+    fn check_ident(
+        &self,
+        chain_id: &ChainId,
+        ident: &ChainIdentifier,
+    ) -> Result<(), IdentValidatorError>;
+}
+
+impl<T: ChainStoreTrait, B: BlockStoreTrait<ChainStore = T>> IdentValidator for B {
+    fn check_ident(
+        &self,
+        chain_id: &ChainId,
+        ident: &ChainIdentifier,
+    ) -> Result<(), IdentValidatorError> {
+        let network_chain = self
+            .chain_store(&chain_id)
+            .ok_or_else(|| IdentValidatorError::UnavailableStore(chain_id.clone()))?;
+        let store_ident = network_chain.chain_identifier();
+
+        if store_ident == &ChainIdentifier::default() {
```

Comment:
This should be something like `store_ident.is_unset()`, and would ideally be implemented by making `ChainIdentifier` an enum that has an `Unset` variant; or even better make `ChainStore.chain_identifier` return a `Result<Option<ChainIdentifier>, Error>`. Relying on the convention that the default means 'not really there' is very brittle.

Reply:
The reason I went with this is basically because the store does not have a way of expressing "unset", it won't be null, since the field is not nullable. This means that to return the Option we'd just hide this assumption further down in the store. The main issue with that IMO is that the store should not know anything about set/unset unless that can be expressed in the data model. 

While this is not strictly true because the store sets the default (since it isn't passed one) I feel like the better place fro all these decisions is in the IdentValidator which would be responsible for all of this. I agree relying on default values is a bit brittle but in this case is a complexity trade-off to avoid a migration in a somewhat critical place.

Reply:
I see two options for addressing that:
1. Change the data model to allow `null` in those columns. That then requires that all code handles the possibility that this is `null`
2. Insist that for new chains we can actually talk to an endpoint to set these fields; if not the chain is ignored like it is today. I think that would be a reasonable requirement, that adding a new chain requires a bit more oversight. It's also not that big a deal if the new chain gets disabled as there can not be any subgraphs for it.

My preference would be (2)

Reply:
I looked into it and I think it's needs to change a bit again because we will need remove the assumption one will always be returned and then expose a way of adding the store. So I think I will do that on a following PR, I feel like this implementation is ok enough as a first step I think it could be merge as is to make the follow update easier to review and get merged ?

Reply:
I decided to give ti a go here since I had some extra time so it's already on the last commit, if you could double-check that it would be great